pycir/cir.py

```diff
@@ -0,0 +1,97 @@
+# -*- coding: utf-8 -*-
+"""
+Centered isotonic regression, generic but more tailored to dose-response and dose-finding data
+
+Assaf Oron, recoded from R 'cir' package
+"""
+
+import numpy as np
+import pandas as pd
+
+
+def cirPAVA(y, x=None, wt=None, outx=None, interiorStrict=True, strict=False, ybounds=None, full=False, dec=False):
+    """
+    Perform regression.
+
+    Args:
+        x,y,wt: vectors of equal length with the doses, mean y values (usually response rates in [0,1]), and weights (often sample size)
+        outx: ...
+    """
+    if ybounds is None:
+        ybounds = np.asarray([0.0,1.0])
+
+    y = np.asarray(y)
+    if y.ndim != 1:
+        raise Exception(f'You supplied an array with {y.ndim} dimensions, but must be a vector (i.e. 1 dimension)')
+
+    m = len(y)
+    dr = pd.DataFrame(data={'x':x,'y':y, 'weight':wt})
+#    if (m <= 1):  ## degenerate case: only one dose level
+#        if (not(full)): return (dr.y)
+#        return(['output':dr,'input':dr,'shrinkage':dr])
+
+    if outx is None:
+        outx = dr.x
+    else:
+        outx = np.asarray(outx)
+
+    dr0 = dr.copy()
+    if dec:
+        dr.loc['y'] = -dr.y
+
+    while True:
+        viol = np.diff(dr.y)<0
+        # This option (default True) makes sure 0s and 1s are not swept into
+        # The shrinkage (in case of binary/binomial observations)
+        if interiorStrict:
+            equals = np.diff(dr.y)==0
+            for i in range(0,m-1):
+                if dr.y[i] in ybounds and dr.y[i+1] in ybounds:
+                    equals[i] = False
+            viol = (viol | equals)
+
+        # strict flag overrides interior-strict nuance
+        if strict:
+            viol = np.diff(dr.y)<=0
+ #       print(viol)
+        if not(any(viol)):
+            break
+
+        i = np.min(np.where(viol==True)) # With sciris instead: i = sc.findfirst(viol)
```

Comment:
If you want to use Sciris this would just be
```
i = sc.findfirst(viol)
```
but probably not worth importing it since this is the only line that's much improved :)